services/rollout-service/pkg/argo/argo.go

```diff
@@ -0,0 +1,341 @@
+}
+
+type ArgoAppProcessor struct {
+	trigger               chan *api.GetOverviewResponse
+	lastOverview          *api.GetOverviewResponse
+	argoApps              chan *v1alpha1.ApplicationWatchEvent
+	ApplicationClient     application.ApplicationServiceClient
+	ManageArgoAppsEnabled bool
+	ManageArgoAppsFilter  []string
+}
+
+func New(appClient application.ApplicationServiceClient, manageArgoApplicationEnabled bool, manageArgoApplicationFilter []string) ArgoAppProcessor {
+	return ArgoAppProcessor{
+		ApplicationClient:     appClient,
+		ManageArgoAppsEnabled: manageArgoApplicationEnabled,
+		ManageArgoAppsFilter:  manageArgoApplicationFilter,
+		trigger:               make(chan *api.GetOverviewResponse),
+		argoApps:              make(chan *v1alpha1.ApplicationWatchEvent),
+	}
+}
+
+type Key struct {
+	AppName     string
+	EnvName     string
+	Application *api.Environment_Application
+	Environment *api.Environment
+}
+
+func (a *ArgoAppProcessor) Push(ctx context.Context, last *api.GetOverviewResponse) {
+	l := logger.FromContext(ctx).With(zap.String("argo-pushing", "ready"))
+	a.lastOverview = last
+	select {
+	case a.trigger <- a.lastOverview:
+		l.Info("argocd.pushed")
+	default:
+	}
+}
+
+func (a *ArgoAppProcessor) Consume(ctx context.Context) error {
+	l := logger.FromContext(ctx).With(zap.String("event-consuming", "ready"))
+
+	for {
+		select {
+		case overview := <-a.trigger:
+			for _, envGroup := range overview.EnvironmentGroups {
+				for _, env := range envGroup.Environments {
+					appList, err := a.ApplicationClient.List(ctx, &application.ApplicationQuery{
+						Name: ptr.FromString(fmt.Sprintf("%s-", env)),
+					})
+
+					if err != nil {
+						fmt.Errorf("listing applications: %w", err)
+					}
+					err = a.DeleteArgoApps(ctx, appList, env.Applications)
+
+					if err != nil {
+						fmt.Errorf("deleting applications: %w", err)
+					}
+
+					for _, app := range env.Applications {
+						a.CreateOrUpdateApp(ctx, overview, app, env)
+					}
+				}
+			}
+
+		case ev := <-a.argoApps:
+			appName, _ := getEnvironmentAndName(ev.Application.Annotations)
+			if appName == "" {
+				continue
+			}
+			switch ev.Type {
+			case "ADDED":
+				l.Info(fmt.Sprintf("argocd.created-%s", appName))
+			case "MODIFIED":
+				l.Info(fmt.Sprintf("argocd.updated-%s", appName))
+			case "DELETED":
+				l.Info(fmt.Sprintf("argocd.deleted-%s", appName))
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (a ArgoAppProcessor) CreateOrUpdateApp(ctx context.Context, overview *api.GetOverviewResponse, app *api.Environment_Application, env *api.Environment) {
+	t := team(overview, app.Name)
+	if a.ManageArgoAppsEnabled && len(a.ManageArgoAppsFilter) > 0 && slices.Contains(a.ManageArgoAppsFilter, t) {
+		k := Key{AppName: app.Name, EnvName: env.Name, Application: app, Environment: env}
+
+		argoApp, err := a.ApplicationClient.Get(ctx, &application.ApplicationQuery{
+			Name: ptr.FromString(app.Name),
```

Comment:
The name is definitely  wrong here as it doesn't contain the environment. Also, why do you even retrieve the application? You did just run a listing operation in the other function which should have given you the exact same object if it exists. Also, when you do the create later,  you still  need to handle the case that the app was already created in the meantime gracefully.

Reply:
Argh, you are right with the duplicate operations. Changed this.

For the graceful handle, not sure how this would be handled.

Reply:
Addressed the handle of the application